tests/migrations/test_operations.py

```diff
@@ -1322,6 +1322,139 @@ def assertIdTypeEqualsFkType():
+        """
+        If RenameField doesn't reload state appropriately, the AlterField
+        crashes on MySQL due to not dropping the PonyRider.pony foreign key
+        constraint before modifying the column.
+        """
+        app_label = 'alter_rename_field_reloads_state_on_fk_target_changes'
+        project_state = self.apply_operations(app_label, ProjectState(), operations=[
+            migrations.CreateModel('Rider', fields=[
+                ('id', models.CharField(primary_key=True, max_length=100)),
+            ]),
+            migrations.CreateModel('Pony', fields=[
+                ('id', models.CharField(primary_key=True, max_length=100)),
+                ('rider', models.ForeignKey('%s.Rider' % app_label, models.CASCADE)),
+            ]),
+            migrations.CreateModel('PonyRider', fields=[
+                ('id', models.AutoField(primary_key=True)),
+                ('pony', models.ForeignKey('%s.Pony' % app_label, models.CASCADE)),
+            ]),
+        ])
+        project_state = self.apply_operations(app_label, project_state, operations=[
+            migrations.RenameField('Rider', 'id', 'id2'),
+            migrations.AlterField('Pony', 'id', models.CharField(primary_key=True, max_length=99)),
+        ])
+
+    @unittest.expectedFailure  # doesn't work due to https://code.djangoproject.com/ticket/25817 (I think)
+    def test_rename_field_reloads_state_on_fk_with_to_field_target_changes(self):
+        """
+        If RenameField doesn't reload state appropriately, the second AlterField
+        crashes on MySQL due to not dropping the PonyRider.slug foreign key
+        constraint before modifying the column.
+        """
+        app_label = 'alter_rename_field_reloads_state_on_fk_with_to_field_target_changes'
+        project_state = self.apply_operations(app_label, ProjectState(), operations=[
+            migrations.CreateModel('Rider', fields=[
+                ('id', models.CharField(primary_key=True, max_length=100)),
+                ('slug', models.CharField(unique=True, max_length=100)),
+            ]),
+            migrations.CreateModel('Pony', fields=[
+                ('id', models.CharField(primary_key=True, max_length=100)),
+                ('rider', models.ForeignKey('%s.Rider' % app_label, models.CASCADE, to_field='slug')),
+                ('slug', models.CharField(unique=True, max_length=100)),
+            ]),
+            migrations.CreateModel('PonyRider', fields=[
+                ('id', models.AutoField(primary_key=True)),
+                ('pony', models.ForeignKey('%s.Pony' % app_label, models.CASCADE, to_field='slug')),
+            ]),
+        ])
+        project_state = self.apply_operations(app_label, project_state, operations=[
+            migrations.RenameField('Rider', 'slug', 'slug2'),
+            migrations.AlterField(
+                'Pony', 'rider', models.ForeignKey('%s.Rider' % app_label, models.CASCADE, to_field='slug2')
+            ),
```

Comment:
FWIW in #9383 this is handled by the `RenameField` operation -- no `AlterField` operation will be generated by the auto-detector just like none are generated on a model and `to` rename.

Reply:
I guess this won't solve the regression in 1.11 since existing migrations would have to be regenerated. On the other hand, no one else reported this issue so hopefully it's not common. Should we cancel plans to push some fix for 1.11 or do you think your patch is suitable?

Reply:
> I guess that won't solve the regression in 1.11 since existing migrations would have to be regenerated.

If we were able to backport #9383 to 1.11 (which is very unlikely given it's blocked by #9389) migrations wouldn't have to be regenerated because the following `AlterField` would be a noop given the [`RenameField` alters the state `to_field`](https://github.com/django/django/pull/9383/files#diff-a71b173d008afea1acd5318ef509bb7cR903).

I'd say let's ship this PR to 1.11 for now and see if we can come up with a non-invasive solution to be backported later.

Reply:
Only to 1.11 or to all branches?

Reply:
I meant all branches.

Reply:
Would you keep the `@unittest.expectedFailure` failure test or not? It sounds like it won't be applicable if your other fix goes through, correct?

Reply:
I'd drop it, and reintroduce the tests in branches where we'll be able to backport #9383.